src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            linkLast(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return node(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = node(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return unlink(node(index));
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> current = first; current != null; current = current.next) {
+                if (current.item == null) {
+                    unlink(current);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> current = first; current != null; current = current.next) {
+                if (object.equals(current.item)) {
+                    unlink(current);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private Node<T> prev;
+        private T item;
+        private Node<T> next;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.prev = prev;
+            this.item = element;
+            this.next = next;
+        }
+    }
+
+    private void linkLast(T value) {
+        Node<T> l = last;
+        Node<T> newNode = new Node<>(l, value, null);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+    }
+
+    private void linkBefore(T value, Node<T> successor) {
+        Node<T> prev = successor.prev;
+        Node<T> newNode = new Node<>(prev, value, successor);
+        successor.prev = newNode;
+        if (prev == null) {
+            first = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        size++;
+    }
+
+    private T unlink(Node<T> target) {
+        final T element = target.item;
+        final Node<T> prev = target.prev;
+        final Node<T> next = target.next;
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            target.prev = null;
+        }
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            target.next = null;
+        }
+        target.item = null;
+        size--;
+        return element;
```

Comment:
definitely, this code works correctly, but I would simplify it because it is difficult to read, but I think in this case the mentor will give better advice

Reply:
private T unlink(Node<T> target) {
        final T element = target.item;
        final Node<T> prev = target.prev;
        final Node<T> next = target.next;
        if (prev == null) {
            first = next;
        } else {
            prev.next = next;
            target.prev = null;
        }
        if (next == null) {
            last = prev;
        } else {
            next.prev = prev;
            target.next = null;
        }
        target.item = null;
        size--;
        return element;
    }

Reply:
I meant this code, which I wrote in the content, because I highlighted a little bit of the wrong area during the review